utils.py

```diff
@@ -0,0 +1,165 @@
+    finally:
+        return None
+
+
+def search_api(query):
+    loop = asyncio.new_event_loop()
+    asyncio.set_event_loop(loop)
+
+    # Query google and big API search
+    google_url = google_search + '&q=%s' % query
+    bing_url = bing_search + '&q=%s' % query
+    req = [{'url': google_url, 'header': None}, {'url': bing_url, 'header': bing_header}]
+    results = loop.run_until_complete(async_main_searches(req))
+
+    # Parse results and combine top 10
+    parsed_results = parse_results(results)
+    top_ten = get_top_ten(parsed_results)
+
+    top_ten['query'] = query
+
+    # Fetch custom snippets based on webpages info
+    loop.run_until_complete(async_main_snippets(top_ten))
+
+    loop.close()
+    return top_ten
+
+
+def parse_results(results):
+    parsed_results = {'googleSearchRes': None, 'googleReqRespTs': None, 'bingSearchRes': None, 'bingReqRespTs': None}
+    if results[0]:
+        parsed_results['googleSearchRes'] = parse_google(results[0])
+        parsed_results['googleReqRespTs'] = results[0]['reqRespTs']
+    if results[1]:
+        parsed_results['bingSearchRes'] = parse_bing(results[1])
+        parsed_results['bingReqRespTs'] = results[1]['reqRespTs']
+    return parsed_results
+
+
+def parse_google(result):
+    parsed = []
+    for item in result['items']:
+        parsed.append({'url': item['link'],
+                       'title': item['title'],
+                       'snippet_orig': item['htmlSnippet'],
+                       'engine': 'Google'})
+    return parsed
+
+
+def parse_bing(result):
+    parsed = []
+    for item in result['webPages']['value']:
+        parsed.append({'url': item['url'],
+                       'title': item['name'],
+                       'snippet_orig': item['snippet'],
+                       'engine': 'Bing'})
+    return parsed
+
+
+def get_top_ten(results):
+    ret = {'searchRes': None, 'googleReqRespTs': None, 'bingReqRespTs': None}
+    if not results['googleSearchRes'] and not results['bingSearchRes']:
+        return None
+    elif not results['googleSearchRes']:
+        ret['searchRes'] = results['bingSearchRes']
+        ret['bingReqRespTs'] = results['bingReqRespTs']
+        return ret
+    elif not results['bingSearchRes']:
+        ret['searchRes'] = results['googleSearchRes']
+        ret['googleReqRespTs'] = results['googleReqRespTs']
+        return ret
+    else:
+        top_ten = []
+        ret['googleReqRespTs'] = results['googleReqRespTs']
+        ret['bingReqRespTs'] = results['bingReqRespTs']
+        google_res = results['googleSearchRes'].copy()
+        bing_res = results['bingSearchRes'].copy()
+        for i in range(10):
+            if results['googleSearchRes'][i]['url'] == results['bingSearchRes'][i]['url']:
+                top_ten.append(results['googleSearchRes'][i])
+                del google_res[i]
+                del bing_res[i]
+
+        google_now = False
+        while len(top_ten) < 10 and google_res and bing_res:
```

Comment:
This round-robin algorithm doesn't prefer pages out of others according to their textual relevance. Better would be to base on textual similarity derived while generating snippets.